Wikipedia/Code/WMFAuthenticationManager.h

```diff
@@ -0,0 +1,86 @@
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef void (^ WMFCaptchaHandler)(NSURL* captchaURL);
+
+/**
+ *  This class manages all aspects of authentication.
+ *
+ *  It abstracts and encapsulates the other fetchers and model classes to provide a
+ *  simple interface for performing authentication tasks.
+ *
+ *  NOTE: This class is NOT STATELESS. Only one authentication operation shoud be running at a given time
+ *  The class will return an error if a second operation is requested while one is in progress
+ *  This behavior will be resolved when the T124408 is completed
+ */
+@interface WMFAuthenticationManager : NSObject
+
+/**
+ *  The current logged in user. If nil, no user is logged in
+ */
+@property (strong, nonatomic, readonly, nullable) NSString* loggedInUsername;
+
+/**
+ *  Returns YES if a user is logged in, NO otherwise
+ */
+@property (assign, nonatomic, readonly) BOOL isLoggedIn;
+
+/**
+ *  Get the shared instance of this class
+ *
+ *  @return The shared Authentication Manager
+ */
++ (instancetype)sharedInstance;
+
+/**
+ *  Get a captcha for account creation. This operation will also check if the username/password is valid.
+ *
+ *  @param username The username to create
+ *  @param password The password for the new user
+ *  @param email    The email for the new user
+ *  @param captcha  The handler for the returned captcha URL
+ *  @param failure  The handler for any errors
+ */
+- (void)getAccountCreationCaptchaWithUsername:(NSString*)username password:(NSString*)password email:(nullable NSString*)email captcha:(WMFCaptchaHandler)captcha failure:(WMFErrorHandler)failure;
```

Comment:
I did `failure: success:` for the PromiseKit changes. Should I change that to `success: failure:` for consistency or should we let it be in whatever order? I'm fine either way. I put success last due to the last param block syntax in swift.